pkg/apis/pipeline/v1beta1/when_types.go

```diff
@@ -0,0 +1,139 @@
+// to determine whether the Task should be executed or skipped
+type WhenExpression struct {
+	// Input is the string for guard checking which can be a static input or an output from a parent Task
+	Input string `json:"input,omitempty"`
+	// Operator that represents an Input's relationship to the values
+	Operator selection.Operator `json:"operator,omitempty"`
+	// Values is an array of strings, which is compared against the input, for guard checking
+	// It must be non-empty
+	Values []string `json:"values,omitempty"`
+}
+
+// GetVarSubstitutionExpressionsForWhenExpression extracts all the values between "$(" and ")"" for a when expression
+func (we *WhenExpression) GetVarSubstitutionExpressionsForWhenExpression() ([]string, bool) {
+	var allExpressions []string
+	allExpressions = append(allExpressions, validateString(we.Input)...)
+	for _, value := range we.Values {
+		allExpressions = append(allExpressions, validateString(value)...)
+	}
+	return allExpressions, len(allExpressions) != 0
+}
+
+func (we *WhenExpression) valuesContainsInput() bool {
+	for i := range we.Values {
+		if we.Values[i] == we.Input {
+			return true
+		}
+	}
+	return false
+}
+
+func (we *WhenExpression) isTrue() bool {
+	if we.Operator == selection.In {
+		return we.valuesContainsInput()
+	}
+	return !we.valuesContainsInput()
+}
+
+func (we *WhenExpression) hasVariable() bool {
+	if _, hasVariable := we.GetVarSubstitutionExpressionsForWhenExpression(); hasVariable {
+		return true
+	}
+	return false
+}
+
+func (we *WhenExpression) applyReplacements(replacements map[string]string) WhenExpression {
+	replacedInput := ApplyReplacements(we.Input, replacements)
+
+	var replacedValues []string
+	for _, val := range we.Values {
+		replacedValues = append(replacedValues, ApplyReplacements(val, replacements))
+	}
+
+	return WhenExpression{Input: replacedInput, Operator: we.Operator, Values: replacedValues}
+}
+
+// WhenExpressions is an array of When Expressions which are used to specify whether a Task should be executed or skipped.
+type WhenExpressions []WhenExpression
+
+// AreTrue indicates whether the When Expression are true.
+func (wes WhenExpressions) AreTrue() bool {
+	for _, we := range wes {
+		if isTrue := we.isTrue(); !isTrue {
+			return false
+		}
+	}
+	return true
+}
+
+// HaveVariables indicates whether When Expressions have variables in the Inputs or Values.
+func (wes WhenExpressions) HaveVariables() bool {
+	for _, we := range wes {
+		if hasVariable := we.hasVariable(); !hasVariable {
+			return false
+		}
+	}
+	return true
+}
+
+// ReplaceWhenExpressionsVariables interpolates variables in When Expressions.
+func (wes WhenExpressions) ReplaceWhenExpressionsVariables(replacements map[string]string) WhenExpressions {
+	var replaced []WhenExpression
+	for _, we := range wes {
+		replaced = append(replaced, we.applyReplacements(replacements))
+	}
+	return replaced
+}
+
+// WhenExpressionEvaluationResult is the result from evaluating a When Expression.
```

Comment:
a general comment about docstrings: its worth including different information than what you get from the name of the thing you are documenting. e.g. in this case, WhenExpressionEvaluationResult kinda already contains the info "is the result from evaluating a when expression"

you could use this docstring to go into more detail, e.g. what does "evaluating" mean exactly? what is the result for?